src/java/org/apache/cassandra/index/sai/plan/QueryController.java

```diff
@@ -158,23 +177,23 @@ public KeyRangeIterator.Builder getIndexQueryResults(Collection<Expression> expr
     {
         KeyRangeIterator.Builder builder = KeyRangeIntersectionIterator.builder(expressions.size());
 
+        Set<Map.Entry<Expression, NavigableSet<SSTableIndex>>> view = referenceAndGetView(expressions).entrySet();
+
         try
         {
-            for (Expression e : expressions)
+            for (Map.Entry<Expression, NavigableSet<SSTableIndex>> e : view)
             {
-                if (e.context.isIndexed())
-                {
-                    @SuppressWarnings({"resource", "RedundantSuppression"}) // RangeIterators are closed by releaseIndexes
-                    KeyRangeIterator memtableIterator = e.context.searchMemtableIndexes(e, mergeRange);
+                @SuppressWarnings({"resource", "RedundantSuppression"}) // RangeIterators are closed by releaseIndexes
+                KeyRangeIterator index = TermIterator.build(e.getKey(), e.getValue(), mergeRange, queryContext);
 
-                    builder.add(memtableIterator);
-                }
+                builder.add(index);
             }
         }
         catch (Throwable t)
         {
             // all sstable indexes in view have been referenced, need to clean up when exception is thrown
             builder.cleanup();
+            view.forEach(e -> e.getValue().forEach(SSTableIndex::release));
```

Comment:
nit: Should we `releaseQuietly()` to guarantee the original `Throwable` is propagated?